src/D2L.CodeStyle.TestAnalyzers/SourceAttribute/TestCaseSourceAttributeAnalyzer.cs

```diff
@@ -0,0 +1,105 @@
+		private const string Category = "Safety";
+
+		private const string Title = "Ensure the source of 'TestCaseSourceAttribute' is static field, property or method in tests.";
+		private const string Description = "The source of 'TestCaseSourceAttribute' must be static field, property or method in tests.";
+		internal const string MessageFormat = "Add 'static' for the source {0} of 'TestCaseSourceAttribute' for NUnit 3 compatibility";
+
+		private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+			DiagnosticId,
+			Title,
+			MessageFormat,
+			Category,
+			DiagnosticSeverity.Warning,
+			isEnabledByDefault: true,
+			description: Description
+		);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create( Rule );
+
+		public override void Initialize( AnalysisContext context ) {
+			context.RegisterSemanticModelAction(
+				AnalyzeSemanticModel
+			);
+		}
+
+		private void AnalyzeSemanticModel( SemanticModelAnalysisContext semanticModelAnalysisContext ) {
+			var tree = semanticModelAnalysisContext.SemanticModel.SyntaxTree;
+			var fields = tree.GetRoot().DescendantNodes().OfType<FieldDeclarationSyntax>().ToImmutableArray();
+			var properties = tree.GetRoot().DescendantNodes().OfType<PropertyDeclarationSyntax>().ToImmutableArray();
+			var methods = tree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().ToImmutableArray();
+
+			// Get All SourceNames of TestCaseSourceAttributes
+			var memberNames = new List<String>();
+			foreach( var method in methods ) {
+				var attributeLists = method.AttributeLists.ToImmutableArray();
+				if( attributeLists.Length == 0 ) {
+					continue;
+				}
+				foreach( var attributeList in attributeLists ) {
+					var attributes = attributeList.Attributes.ToImmutableArray();
+					foreach( var attribute in attributes ) {
+						if( attribute.Name.ToString() == "TestCaseSource" ) {
+							var attributeArguments = attribute.ArgumentList.Arguments.ToImmutableArray();
+							foreach( var attributeArgument in attributeArguments ) {
+								var expression = attributeArgument.Expression;
+								//StringLiteralExpression: "a", invocationExpression: nameof(a)  
+								if( expression is LiteralExpressionSyntax ) {
+									var member = expression.ToString().Replace( "\"", "" ).Replace( "\'", "" ).Trim();
+									memberNames.Add( member );
+								} else if( expression is InvocationExpressionSyntax ) {
+									var member = expression.ToString().Trim().Substring( 7, expression.ToString().Trim().Length - 8 ).Trim();
+									memberNames.Add( member );
+								}
+							}
+						}
+					}
+				}
+			}
+
+			// For each Source Report Diagnostic
+			foreach( var method in methods ) {
+				if( memberNames.Contains( method.Identifier.ToString() ) ) {
+					if( !method.Modifiers.Any( SyntaxKind.StaticKeyword ) ) {
+						var diagnostic = Diagnostic.Create( Rule, method.GetLocation(), "method (and the called methods by it)" );

```

Comment:
Is it a better idea to report the diagnostics at the usages of the attribute instead?
I.e., something like this:

```csharp
private readonly int[] items = { 1, 2, 3, 4 };

[Test, TestCaseSource( nameof(item) )] // this line errors out, saying "'items' need to be static."
public void Test( int item ) { }
```